{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SearchIO\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "from Bio.Seq import Seq\n",
    "from Bio import AlignIO\n",
    "\n",
    "from io import StringIO\n",
    "import subprocess\n",
    "\n",
    "from os import listdir\n",
    "from os.path import isfile, join, dirname, isdir, exists\n",
    "from os import path\n",
    "import os\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#These are the queries for the hmmsearch command. Change the genus or domain to your liking\n",
    "\n",
    "genus = 'magnaporthe' #or try 'neurospora'\n",
    "domain = 'NACHT' #try 'NACHT', 'NB-ARC', or 'HET'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#helper function for preparing directories. do not change this.\n",
    "\n",
    "def make_path(*argv):\n",
    "    mypath = path.join(*argv)\n",
    "    if not path.exists(dirname(mypath)):\n",
    "        os.mkdir(dirname(mypath))\n",
    "    return mypath\n",
    "\n",
    "def make_dir(*argv):\n",
    "    mydir = path.join(*argv)\n",
    "    if not path.exists(mydir):\n",
    "        os.mkdir(mydir)\n",
    "    return mydir\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#preparing path names for hmmsearch command...\n",
    "proteome_path = make_path('proteomes', genus, os.listdir(make_dir('proteomes', genus))[0])\n",
    "genomes_dir = make_dir('genomes', genus)\n",
    "assembly_name = proteome_path.split('_')[-2]\n",
    "domain_path = make_path('pfam_domains', domain+'.hmm')\n",
    "hmmsearch_align_path = make_path('hmmsearch_align', genus, domain+'-'+assembly_name+'.aln')\n",
    "hmmsearch_out_path = make_path('hmmsearch_out', genus, domain+'-'+assembly_name+'.out')\n",
    "hmmmsearch_domout_path = make_path('hmmsearch_out', genus, domain+'-'+assembly_name+'.domout')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "!hmmsearch --domE 0.01  -A $hmmsearch_align_path --domtblout $hmmmsearch_domout_path $domain_path $proteome_path > $hmmsearch_out_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E-value: 1e-13 for XP_003720097.1 uncharacterized protein MGG_17753 [Pyricularia oryzae 70-15]\n",
      "E-value: 1.3e-11 for XP_003711083.1 ankyrin repeatl protein [Pyricularia oryzae 70-15]\n",
      "E-value: 9.9e-10 for XP_003718688.1 uncharacterized protein MGG_00388 [Pyricularia oryzae 70-15]\n",
      "E-value: 3.1e-09 for XP_003712653.1 uncharacterized protein MGG_05125 [Pyricularia oryzae 70-15]\n",
      "E-value: 3.5e-08 for XP_003712408.1 uncharacterized protein MGG_04917 [Pyricularia oryzae 70-15]\n",
      "E-value: 3.3e-09 for XP_003713330.1 uncharacterized protein MGG_15509 [Pyricularia oryzae 70-15]\n",
      "E-value: 1.5e-08 for XP_003713943.1 uncharacterized protein MGG_08914 [Pyricularia oryzae 70-15]\n",
      "E-value: 3.3e-08 for XP_003720118.1 uncharacterized protein MGG_12047 [Pyricularia oryzae 70-15]\n",
      "E-value: 2.7e-08 for XP_003710235.1 ankyrin repeat domain-containing protein 29 [Pyricularia oryzae 70-15]\n",
      "E-value: 9.4e-08 for XP_003721297.1 NACHT and Ankyrin domain-containing protein [Pyricularia oryzae 70-15]\n",
      "E-value: 2.5e-07 for XP_003716052.1 uncharacterized protein MGG_08542 [Pyricularia oryzae 70-15]\n",
      "E-value: 8.8e-07 for XP_003721407.1 WD domain-containing protein [Pyricularia oryzae 70-15]\n",
      "E-value: 3.1e-06 for XP_003720270.1 NACHT domain-containing protein [Pyricularia oryzae 70-15]\n",
      "E-value: 2e-06 for XP_003710517.1 uncharacterized protein MGG_16427 [Pyricularia oryzae 70-15]\n",
      "E-value: 6.6e-06 for XP_003714937.1 uncharacterized protein MGG_07985 [Pyricularia oryzae 70-15]\n",
      "E-value: 4.6e-06 for XP_003711012.1 uncharacterized protein MGG_04406 [Pyricularia oryzae 70-15]\n",
      "E-value: 5.5e-06 for XP_003716910.1 uncharacterized protein MGG_06617 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0055 for XP_003709948.1 peroxisome biosynthesis protein [Pyricularia oryzae 70-15]\n",
      "E-value: 9.4e-06 for XP_003709501.1 NACHT and WD domain-containing protein [Pyricularia oryzae 70-15]\n",
      "E-value: 9.5e-06 for XP_003718480.1 uncharacterized protein MGG_00567 [Pyricularia oryzae 70-15]\n",
      "E-value: 3.4e-05 for XP_003714457.1 origin recognition complex subunit 4 [Pyricularia oryzae 70-15]\n",
      "E-value: 2e-05 for XP_003711680.1 ankyrin repeat and SOCS box protein 7 [Pyricularia oryzae 70-15]\n",
      "E-value: 4.1e-05 for XP_003712649.1 uncharacterized protein MGG_05121 [Pyricularia oryzae 70-15]\n",
      "E-value: 4.1e-05 for XP_003708751.1 uncharacterized protein MGG_15994 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0091 for XP_003714400.1 uncharacterized protein MGG_15701 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.00028 for XP_003720645.1 uncharacterized protein MGG_03038 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.00073 for XP_016846048.1 uncharacterized protein MGG_10236 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.00083 for XP_003714909.1 uncharacterized protein MGG_15330 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0016 for XP_003713541.1 thermosensitive gluconokinase [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0013 for XP_003712089.1 GTP binding protein Bms1 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0019 for XP_003720724.1 uncharacterized protein MGG_02976 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0026 for XP_003712027.1 uncharacterized protein MGG_06149 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0041 for XP_003718065.1 uncharacterized protein MGG_17348 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.0075 for XP_003713691.1 uncharacterized protein MGG_15566 [Pyricularia oryzae 70-15]\n",
      "E-value: 0.009 for XP_003719038.1 ATPase NPA3 [Pyricularia oryzae 70-15]\n"
     ]
    }
   ],
   "source": [
    "#Parse hmmsearch output and filter hits by e-value. Find these hits the reference proteome and display annotations of the numTop proteins along with their E-values\n",
    "\n",
    "numTop = 35 ## change this to adjust the number of protein descriptions displayed\n",
    "ethresh = 0.01 ### adjust this for e-value threshold\n",
    "\n",
    "qresult = next(SearchIO.parse(hmmmsearch_domout_path, 'hmmsearch3-domtab'))\n",
    "filtered_hits = list(filter(lambda hsp: hsp.evalue < ethresh, qresult.hsps))\n",
    "hit_ids = [hsp.hit_id for hsp in filtered_hits]\n",
    "hit_sequences = sorted([record for record in SeqIO.parse(proteome_path, \"fasta\") if record.id in hit_ids], key=lambda a: hit_ids.index(a.id))\n",
    "print('\\n'.join([f'E-value: {filtered_hits[ii].evalue} for ' + hit_sequences[ii].description for ii in range(numTop)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Exonerates a protein to all genomes for the species. don't change this.\n",
    "def exonerate(nlr_protein):    \n",
    "    nlr_path = make_path('nlr_protein', genus, 'nlr_'+nlr_protein.id+'.fasta')\n",
    "    p2gaa_path = make_path('exonerate_p2gaa', genus, 'p2gaa_%s.faa' % nlr_protein.id)\n",
    "    p2gnt_path = make_path('exonerate_p2gnt', genus, 'p2gnt_%s.fasta' % nlr_protein.id)\n",
    "    SeqIO.write(nlr_protein, nlr_path, \"fasta\")\n",
    "    \n",
    "    files = [f for f in listdir(genomes_dir) if isfile(join(genomes_dir, f)) and f != '.DS_Store' ]\n",
    "    numExonerated = 0\n",
    "\n",
    "    for ii, f in enumerate(files):\n",
    "        proc = subprocess.run(['exonerate', '--model', 'protein2genome', '--bestn', '1', '--showvulgar', 'F', '--ryo', '''\">%ti (%tab - %tae)\\n%tas\\n\"''', nlr_path, join(genomes_dir, f)],  stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "\n",
    "        outsplit = proc.stdout.decode().split('''\"''')\n",
    "        if len(outsplit) > 3:\n",
    "#             print('exonerating', f, 'counter: ', ii)\n",
    "            numExonerated += 1\n",
    "\n",
    "            qresult = next(SearchIO.parse(StringIO(outsplit[-3]), 'exonerate-text'))\n",
    "            aa_record = sum([frag.hit for frag in qresult[0][0]], Seq(''))\n",
    "\n",
    "            if not ii:\n",
    "                writeMode = 'w'\n",
    "            else:\n",
    "                writeMode = 'a'\n",
    "            \n",
    "            aafile = open(p2gaa_path, writeMode)        \n",
    "            SeqIO.write(aa_record, aafile, 'fasta')\n",
    "            aafile.close()\n",
    "\n",
    "            ntfile = open(p2gnt_path, writeMode)        \n",
    "            ntseq = outsplit[-2]\n",
    "            ntfile.write(ntseq)\n",
    "            ntfile.close()\n",
    "            \n",
    "            if not ii % 50:\n",
    "                print(ii+1, 'sequences exonerated')\n",
    "#         else:\n",
    "#             print('failed to exonerate', f, 'counter: ', ii)\n",
    "\n",
    "    return numExonerated, ii+1\n",
    "\n",
    "\n",
    "#get column of position weight matrix\n",
    "def get_distr(aa_col, pseudocounts = 0.0):\n",
    "    abundict = {aa:pseudocounts for aa in list('ACDEFGHIKLMNPQRSTVWY-X*')}\n",
    "    for site in aa_col:\n",
    "        if site in abundict.keys():\n",
    "            abundict[site] += 1.0\n",
    "        else:\n",
    "            print('did not find', site, 'in alphabet')\n",
    "            raise Exception()\n",
    "    abundict = {aa:count for aa,count in abundict.items() if count > 0.0}\n",
    "    weights = np.array(list(abundict.values()))\n",
    "    distr = weights/sum(weights)\n",
    "    return distr\n",
    "\n",
    "#compute entropy from a list of weights\n",
    "def get_entropy(distr):\n",
    "    return -np.sum(np.log(distr)*distr)\n",
    "\n",
    "#returns entropy sequence from multiple alignment\n",
    "def get_entropy_sequence(align):\n",
    "    entropies = []\n",
    "    for col in range(len(align[0,:])):\n",
    "        entropies.append(get_entropy(get_distr(align[:, col])))\n",
    "    return np.array(entropies)\n",
    "\n",
    "#plot entropy sequence\n",
    "def plot_entropy(entropies, protein_id):\n",
    "    plt.plot(entropies)\n",
    "    plt.xlabel('Amino acid position')\n",
    "    plt.ylabel('Entropy')\n",
    "    plt.title('Splice-aware entropy of %s' % protein_id)\n",
    "    \n",
    "    \n",
    "# def save_plot(nlr_protein):    \n",
    "#     entropy_seq_path = make_path('entropy', genus, 'seq', 'entropy_clustalo_%s.npy' % nlr_protein.id)\n",
    "#     plot_entropy(np.load(entropy_seq_path), nlr_protein.id)\n",
    "#     entropy_plot_path = make_path('entropy', genus, 'plot', 'entropy_clustalo_%s.png' % nlr_protein.id)\n",
    "#     plt.savefig(entropy_plot_path)    \n",
    "    \n",
    "def get_multiple_alignment(nlr_protein):\n",
    "    print('exonerating protein', nlr_protein.id, '...')\n",
    "    numExonerated, numGenomes = exonerate(nlr_protein)\n",
    "    print(numExonerated, 'genomes out of', numGenomes, 'total contain an ortholog of protein', nlr_protein)\n",
    "    \n",
    "    p2gaa_path = make_path('exonerate_p2gaa', genus, 'p2gaa_%s.faa' % nlr_protein.id)\n",
    "    clustalo_out_path = make_path('clustalo_out', genus, 'clustalo_%s.out' % nlr_protein.id)\n",
    "    !clustalo -i $p2gaa_path -o $clustalo_out_path\n",
    "    \n",
    "def save_entropy(nlr_protein):\n",
    "    clustalo_out_path = make_path('clustalo_out', genus, 'clustalo_%s.out' % nlr_protein.id)\n",
    "    entropies = get_entropy_sequence(AlignIO.read(clustalo_out_path, \"fasta\"))\n",
    "    entropy_seq_path = make_path('entropy', genus, 'seq', 'entropy_clustalo_%s.npy' % nlr_protein.id)    \n",
    "    np.save(entropy_seq_path, entropies)\n",
    "    entropy_plot_path = make_path('entropy', genus, 'plot', 'entropy_clustalo_%s.png' % nlr_protein.id)\n",
    "    plot_entropy(entropies, nlr_protein.id)\n",
    "    plt.savefig(entropy_plot_path)\n",
    "    plt.close()\n",
    "\n",
    "def main(nlr_protein):\n",
    "    get_multiple_alignment(nlr_protein)\n",
    "    save_entropy(nlr_protein)\n",
    "    \n",
    "def loop_over_hits(hits):\n",
    "    for nlr_protein in hits:\n",
    "        main(nlr_protein)\n",
    "    print('completed pipeline on protein', nlr_protein.id)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exonerating protein XP_003712653.1 ...\n",
      "1 sequences exonerated\n",
      "51 sequences exonerated\n",
      "101 sequences exonerated\n",
      "151 sequences exonerated\n",
      "201 sequences exonerated\n",
      "250 genomes out of 250 total contain an ortholog of protein ID: XP_003712653.1\n",
      "Name: XP_003712653.1\n",
      "Description: XP_003712653.1 uncharacterized protein MGG_05125 [Pyricularia oryzae 70-15]\n",
      "Number of features: 0\n",
      "Seq('MPSNSEYTIGWIAAIETEFVAASRFLDKRHGRPESQDSADSNSYKLGSMEGHNV...RFV')\n",
      "FATAL: Cowardly refusing to overwrite already existing file 'clustalo_out/magnaporthe/clustalo_XP_003712653.1.out'. Use --force to force overwriting.\n"
     ]
    }
   ],
   "source": [
    "# use the main function to exonerate a protein against the genomes,\n",
    "# compute multiple alignment, and save entropy plots\n",
    "# also try loop_over_hits(hit_sequences) to run the pipeline on\n",
    "# a list hit sequences found by hmmer\n",
    "\n",
    "main(hit_sequences[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
